src/main/scala/com/microsoft/hyperspace/index/dataskipping/DataSkippingIndex.scala

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (2021) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.index.dataskipping
+
+import org.apache.spark.sql.{Column, DataFrame, SaveMode}
+import org.apache.spark.sql.functions.{input_file_name, min, spark_partition_id}
+
+import com.microsoft.hyperspace.index._
+import com.microsoft.hyperspace.index.dataskipping.sketch.Sketch
+import com.microsoft.hyperspace.index.dataskipping.util.ExpressionUtils
+import com.microsoft.hyperspace.util.HyperspaceConf
+
+/**
+ * DataSkippingIndex is an index that can accelerate queries by filtering out
+ * files in relations using sketches.
+ *
+ * @param sketches List of sketches for this index
+ * @param properties Properties for this index; see [[Index.properties]] for details.
+ */
+case class DataSkippingIndex(
+    sketches: Seq[Sketch],
+    override val properties: Map[String, String] = Map.empty)
+    extends Index {
+  assert(sketches.nonEmpty, "At least one sketch is required.")
+
+  @transient
+  lazy val aggregateFunctions = sketches.flatMap { s =>
+    val aggrs = s.aggregateFunctions
+    assert(aggrs.nonEmpty)
+    aggrs.zipWithIndex.map {
+      case (aggr, idx) =>
+        new Column(aggr).as(getNormalizeColumnName(s"${s}_$idx"))
+    }
+  }
+
+  override def kind: String = DataSkippingIndex.kind
+
+  override def kindAbbr: String = DataSkippingIndex.kindAbbr
+
+  override def indexedColumns: Seq[String] = sketches.flatMap(_.indexedColumns).distinct
+
+  override def referencedColumns: Seq[String] = sketches.flatMap(_.referencedColumns).distinct
+
+  override def withNewProperties(newProperties: Map[String, String]): DataSkippingIndex = {
+    copy(properties = newProperties)
+  }
+
+  override def statistics(extended: Boolean = false): Map[String, String] = {
+    Map("sketches" -> sketches.mkString(", "))
+  }
+
+  override def canHandleDeletedFiles: Boolean = true
+
+  override def write(ctx: IndexerContext, indexData: DataFrame): Unit = {
```

Comment:
Can we have a better naming on this?
This is for create index & refresh full, but seems write is not intuitive compared to refreshIncremental / optimize.
WDYT?

Reply:
In the Scaladoc for `Index.write` I've written:

> Writes the index data to the specified path.

and it is precisely what it is supposed to do. We can use other words like "save", "put", or "store". But I think "write" isn't bad. "save" also sounds good.